substrate/frame/contracts/proc-macro/src/lib.rs

```diff
@@ -206,9 +207,25 @@ impl HostFn {
 					}
 					not_deprecated = false;
 				},
+				"mutable" => {
+					if mutable {
+						return Err(err(span, "#[mutable] can only be specified once"))
+					}
+					mutable = true;
+				},
 				_ => return Err(err(span, msg)),
 			}
 		}
+
+		if mutable {
+			let stmt = syn::parse_quote! {
+				if ctx.ext().is_read_only() {
+					return Err(Error::<E::T>::StateChangeDenied.into());
+				}
+			};
+			item.block.stmts.insert(0, stmt);
+		}
```

Comment:
This is dangerous because we error out without charging any gas. Even when doing no work there is an overhead of calling a host function which needs to be charged.  We need to merge https://github.com/paritytech/polkadot-sdk/pull/4233 first. And then make sure that we do things in this order:

1. Charge the general host function overhead.
2. Do the check added in this PR.
3. Charge the host function specific gas.

Reply:
1. should be done in #4233
https://github.com/paritytech/polkadot-sdk/blob/pg/rework-host-benchs/substrate/frame/contracts/proc-macro/src/lib.rs?plain=1#L708-L711

> 3. Charge the host function specific gas.

What else should you charge if you exited early ?

Reply:
I don't understand your question regarding 3. But generally, we don't refund on error. But we can make an exception for the read_only check.

Reply:
I was just wondering what you meant by " charge the host function specific gas ", we would just exit early here so I don't see how you could charge anything else.

Reply:
Point 1. + 2. will be done in the macro. Point 3.  has to be done manually as we already do.